### ISSUE TITLE
Support for listwise loss functions in bi-encoder

### ISSUE THREAD

**Issue by madhumitha-mohan** [UKPLab/sentence-transformers#3492]
Hello! 
I am currently working on a retrieval system using bi-encoder. The loss function I use right now is [MultipleNegativesRankingLoss](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#multiplenegativesrankingloss). This loss function currently supports <query, positive, negative> triplets with in-batch negatives. 

For my use-case I have a graded relevance metric, where, for a given query I have multiple documents each with different grades of relevance (say in scale 0 to 5). 
- Is there a listwise loss function I can use for bi-encoder? 
- Which existing loss function in sentence-transformers is best suited for this use case? 
- How can we modify existing loss functions to adapt to this use case? For instance adding weights on MultipleNegativesRankingLoss to support graded relevance labels

Thank you!

**Comment by kudos07** [comment]
Hey, thanks for the question!

**1. Listwise loss**  
Sentence-transformers doesn’t include a true listwise loss (e.g., ListNet, ListMLE, ApproxNDCG). Most losses are pairwise or pointwise. You can implement a custom `SentenceTransformerLoss` with `(query, docs, relevance_scores)` and compute a listwise objective (e.g., ApproxNDCG, ListNet).

**2. Best existing option for graded relevance**  
`CosineSimilarityLoss` is the most practical since it supports continuous labels (normalize 0–5 → [0,1]).  
- `SoftmaxLoss`: possible, but loses the ordinal structure.  
- `MultipleNegativesRankingLoss`: strong for binary relevance, less suited for graded.

**3. Adapting losses**  
You can extend `MultipleNegativesRankingLoss` into a weighted version:  
- w = rel_score / max_rel  
- Scale each pair’s loss by w  

This emphasizes highly relevant docs while still learning from weaker ones. Alternatively, use a listwise softmax with normalized relevance scores as the target distribution.

**Comment by ShreyGanatra** [comment]
What if I want to attend to all documents at once like input is (query [SEP] doc1 [SEP] doc2 ...) instead of
```
# Create input pairs for the model
pairs = [(query, document) for query, docs in zip(queries, docs_list) for document in docs]
```
Can this be done without writing custom loss?
